File: epi_judge_python/is_tree_balanced.py
from binary_tree_node import BinaryTreeNode
from test_framework import generic_test



def is_balanced_binary_tree(tree: BinaryTreeNode) -> bool:

    # def height(tree: BinaryTreeNode):
    #     if tree is None:
    #         return 0
        
    #     return max(height(tree.left),height(tree.right))+1


    def is_balanced(tree: BinaryTreeNode):
        
        # a tree is balanced if its left subtree and right subtree are balanced
        if tree is None:
            return (True,0)

        (leftBalance,leftHeight)=is_balanced(tree.left)
        # should do early termination here because if leftBalance is false then we can quit without
        # looking at the right side 
        (rightBalance,rightHeight)=is_balanced(tree.right)
        retval=True

        if (abs(leftHeight-rightHeight) > 1):
            retval=False
        if leftBalance is False:
            retval=False
        if rightBalance is False:
            retval=False
        
        myheight=max(leftHeight,rightHeight)+1
        return (retval,myheight)


    return is_balanced(tree)[0]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('is_tree_balanced.py',
                                       'is_tree_balanced.tsv',
                                       is_balanced_binary_tree))

File: epi_judge_cpp/is_tree_balanced.cc
#include "binary_tree_node.h"
#include "test_framework/generic_test.h"


std::pair<int,bool> IsBalancedRec(const unique_ptr<BinaryTreeNode<int>>& tree){

  if(tree==nullptr){
      return std::make_pair(0,true);
    }

    auto leftPack=IsBalancedRec(tree->left);
    auto rightPack=IsBalancedRec(tree->right);
    if (leftPack.second==false || rightPack.second==false){
      return std::make_pair(0,false);
    }
    if(abs(leftPack.first-rightPack.first) > 1){
      return std::make_pair(0,false);
    }

    int height = std::max(leftPack.first,rightPack.first)+1;
    return std::make_pair(height,true);

}

bool IsBalanced(const unique_ptr<BinaryTreeNode<int>>& tree) {
  
  
  return IsBalancedRec(tree).second;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"tree"};
  return GenericTestMain(args, "is_tree_balanced.cc", "is_tree_balanced.tsv",
                         &IsBalanced, DefaultComparator{}, param_names);
}
